File: src/main/java/org/kite9/framework/logging/Kite9Log.java
package org.kite9.framework.logging;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Very simple abstraction class for all logging functions.  
 * 
 * @author robmoffat
 *
 */
public class Kite9Log {

	Logable logFor;
	
	static boolean logging = isLoggingOn();

	static PrintStream logFile;
	
	public static final boolean OUTPUT_TO_LOG = false;
	
	static {
		if (OUTPUT_TO_LOG) {
			try {
				logFile = new PrintStream(new FileOutputStream(new File("kite9.log")));
			} catch (FileNotFoundException e) {
				logFile = System.out;
			}
		} else {
			logFile = System.out;
		}

	}
	
	private static boolean isLoggingOn() {
		String propVal = System.getProperty("kite9.logging");
		if (propVal==null) {
			return true;
		}
		
		propVal=propVal.toLowerCase().trim();
		
		if ("on".equals(propVal) || "true".equals(propVal) || "yes".equals(propVal)) {
			return true;
		} else {
			return false;
		}
	}
	
	public static void setLogging(boolean state) {
		logging = state;
	}
	
	public boolean go() {
		return !logging;
	}

	public Kite9Log(Logable o) {
		this.logFor = o;
	}

	public void send(String string) {
		if (logFor.isLoggingEnabled() && logging)
			logFile.println(logFor.getPrefix() + " " + string);
	}

	public void send(String prefix, Collection<?> items) {
		if (logFor.isLoggingEnabled() && logging) {
			logFile.println(logFor.getPrefix() + " " + prefix);

			StringBuffer sb = new StringBuffer();
			for (Object o : items) {
				sb.append("\t");
				sb.append(o.toString());
				sb.append("\n");
			}

			logFile.println(sb.toString());

		}
	}

	public void send(String prefix, Table t) {
		if (logFor.isLoggingEnabled() && logging) {
			StringBuffer sb = new StringBuffer(1000);
			t.display(sb);
			send(prefix + sb.toString());
		}
	}

	public void send(String prefix, Map<?, ?> items) {
		if (logFor.isLoggingEnabled() && logging) {
			logFile.println(logFor.getPrefix() + " " + prefix);
			Table t = new Table();
			Set<?> keys = items.keySet();
			List<Object> keyList = new ArrayList<Object>(keys);
			Collections.sort(keyList, new Comparator<Object>() {

				public int compare(Object o1, Object o2) {
					if (o1 == null) {
						return -1; 
					} else if (o2 == null) {
						return 1;
					} else {
						return o1.toString().compareTo(o2.toString());
					}
				}
				
			});

			for (Object object : keyList) {
				if (object != null) {
					t.addRow(new Object[] { "\t", object.toString(), items.get(object).toString() });
				}
			}

			StringBuffer sb = new StringBuffer();
			t.display(sb);
			logFile.println(sb.toString());
		}
	}

	public void error(String string) {
	    System.err.println(logFor.getPrefix()+" "+string);
	}

	public void error(String string, Throwable e) {
	    System.err.println(logFor.getPrefix()+" "+string);
	    e.printStackTrace();
	}
}

File: src/main/kotlin/org/kite9/diagram/logging/Kite9Log.kt
package org.kite9.diagram.logging

import org.kite9.diagram.logging.Logable
import org.kite9.diagram.logging.Kite9Log
import java.io.File
import java.lang.StringBuffer
import java.util.Collections
import java.io.PrintStream
import java.io.FileOutputStream
import java.io.FileNotFoundException
import java.lang.Exception
import java.util.ArrayList

/**
 * Very simple abstraction class for all logging functions.
 *
 * @author robmoffat
 */
class Kite9Log(var logFor: Logable) {
    enum class Destination {
        OFF, STREAM, FILE
    }

    fun go(): Boolean {
        return logFile == null
    }

    fun send(string: String?) {
        if (logFor.isLoggingEnabled && logFile != null) logFile!!.println(logFor.prefix + " " + string)
    }

    fun send(indent: Int, string: String?) {
        if (logFor.isLoggingEnabled && logFile != null) logFile!!.print(logFor.prefix)
        logFile!!.write(INDENT.toByteArray(), 0, 1 + indent)
        logFile!!.println(string)
    }

    fun send(prefix: String?, items: Collection<*>) {
        if (logFor.isLoggingEnabled && logFile != null) {
            logFile!!.println(logFor.prefix + " " + prefix)
            val sb = StringBuffer()
            for (o in items) {
                sb.append("\t")
                sb.append(o.toString())
                sb.append("\n")
            }
            logFile!!.println(sb.toString())
        }
    }

    fun send(prefix: String?, t: Table) {
        if (logFor.isLoggingEnabled && logFile != null) {
            val sb = StringBuffer(1000)
            t.display(sb)
            send(prefix + sb.toString())
        }
    }

    fun send(prefix: String?, items: Map<*, *>) {
        if (logFor.isLoggingEnabled && logFile != null) {
            logFile!!.println(logFor.prefix + " " + prefix)
            val t = Table()
            val keys = items.keys
            val keyList: ArrayList<Any?> = ArrayList(keys)
            Collections.sort(keyList) { o1, o2 ->
                if (o1 == null) {
                    -1
                } else if (o2 == null) {
                    1
                } else {
                    o1.toString().compareTo(o2.toString())
                }
            }
            for (`object` in keyList) {
                if (`object` != null) {
                    t.addRow(*arrayOf<Any>("\t", `object`.toString(), items[`object`].toString()))
                }
            }
            val sb = StringBuffer()
            t.display(sb)
            logFile!!.println(sb.toString())
        }
    }

    fun error(string: String?) {
        System.err.println(logFor.prefix + " " + string)
    }

    fun error(string: String?, e: Throwable) {
        System.err.println(logFor.prefix + " " + string)
        e.printStackTrace()
    }

    fun send(prefix: String?, arg0: Exception) {
        logFile!!.println(logFor.prefix + " " + prefix)
        arg0.printStackTrace(System.out)
    }

    companion object {
        private val INDENT = String(CharArray(100)).replace('\u0000', ' ')
        var logFile: PrintStream? = null
        fun setLogging(state: Destination) {
            if (state == Destination.FILE) {
                try {
                    logFile = PrintStream(FileOutputStream(File("kite9.log")))
                } catch (e: FileNotFoundException) {
                    e.printStackTrace()
                    logFile = System.out
                }
            } else if (state == Destination.STREAM) {
                logFile = System.out
            } else {
                logFile = null
            }
        }
    }
}